pennylane_lightning/src/Gates.hpp

```diff
@@ -323,6 +323,23 @@ namespace Pennylane {
         void applyKernel(const StateVector& state, const std::vector<size_t>& indices, const std::vector<size_t>& externalIndices);
     };
 
+    class Unitary : public AbstractGate {
+    private:
+        const std::vector<CplxType> matrix;
```

Comment:
We actually don't explicitly define `matrix` and so `matrix` and `asMatrix` are not needed. However, they need to be defined in the header because they are virtual in `AbstractGate`.

Reply:
> If we have params as a 2*d length flattened and real vector, can we convert that to a d-length complex vector with no overhead/copying?

This solution already involves copying (the `Unitary` constructor is calling the `vector` copy constructor). In fact, because we're relying on the numpy implicit conversion from a python list `op_param` to `vector<vector<double>> params`, the overhead is already awful, scaling with the size of the matrix! So we really should be passing it in separately.

Reply:
> because we're relying on the pybind implicit conversion from a python list op_param to vector<vector<double>> params, the overhead is already awful, scaling with the size of the matrix

Actually I take this back; this may not be the case. Its a python list of what's potentially a mixture of python lists and numpy array views. I'm not entirely sure how that gets handled.